### PR TITLE
feat(querylog): Introduce querylog dataset and processor

### PR DESCRIPTION
**Pull request by lynnagara**
The querylog dataset will store metadata about each query run on Snuba.

The querylog_dist table will need to be created before this is merged.

### CODE REVIEW

File: snuba/datasets/querylog.py
Comment (lynnagara) [inline snuba/datasets/querylog.py]: Not sure if this is really useful, but it seems like it won't hurt to support our standard timeseries queries on this dataset.

File: snuba/datasets/querylog.py
Comment (fpacifici) [inline snuba/datasets/querylog.py]: I can see us adding columns (and then dropping them) relatively often to track specific issues, I don't think we can afford to change the schema each time, so (to be discussed with the team) but I think a concept of tags to add/remove new fields quickly would be useful.

Some example I can think we will want to add:
- serialize the AST for debug
- list the query processors we applied 
- write a more structured representation of the query to be able to do some breakdowns.


Also we likely need trace id and span id to correlate with APM data.

File: snuba/datasets/querylog.py
Comment (fpacifici) [inline snuba/datasets/querylog.py]: I would add the timing columns in the clickhouse query as well even if you are not populating them yet, we know we will need them soon, so we avoid spending the time for the migration.

File: snuba/datasets/querylog.py
Comment (markstory) [inline snuba/datasets/querylog.py]: Do you want the organization if available?

File: snuba/datasets/querylog.py
Comment (lynnagara) [reply snuba/datasets/querylog.py]: Added duration_ms to the nested queries (I don't think anything else is particuarly useful)

File: snuba/datasets/querylog.py
Comment (lynnagara) [reply snuba/datasets/querylog.py]: Yeah, will add the field (even if not yet populated)

File: snuba/datasets/querylog.py
Comment (lynnagara) [reply snuba/datasets/querylog.py]: Yeah, that makes sense especially for ad hoc dataset specific stuff that might not be relevant for all logged queries. I'd rather add that when we actually need it though, I don't think adding tags later when we confirm that we need and want to go down that route will be too difficult.

File: snuba/datasets/querylog.py
Comment (fpacifici) [inline snuba/datasets/querylog.py]: We generally leave a blank line between system imports and snuba imports, so here and not at line 17

File: snuba/datasets/querylog.py
Comment (fpacifici) [inline snuba/datasets/querylog.py]: out of order

File: snuba/datasets/querylog.py
Comment (fpacifici) [inline snuba/datasets/querylog.py]: please add a sample clause, just in case, because we cannot add it later.

File: snuba/datasets/querylog.py
Comment (fpacifici) [inline snuba/datasets/querylog.py]: I guess this will be more useful as toDate(timestamp)

File: snuba/datasets/querylog.py
Comment (fpacifici) [inline snuba/datasets/querylog.py]: Why not adding the full stats field ? There are a few other interesting fields inside like whether the query is consistent, the results size, data about the rate limiting, etc.

File: snuba/datasets/querylog.py
Comment (fpacifici) [inline snuba/datasets/querylog.py]: there is no start_ts in this table.

File: snuba/datasets/querylog.py
Comment (fpacifici) [inline snuba/datasets/querylog.py]: There is no project_id column, I think queries will not work this way. More in general I don;t think you need the project extension at all, we use it in other datasets to enforce queries are done by project.

File: snuba/datasets/querylog_processor.py
Comment (fpacifici) [inline snuba/datasets/querylog_processor.py]: __get_request

File: snuba/datasets/querylog_processor.py
Comment (fpacifici) [inline snuba/datasets/querylog_processor.py]: same: __get_sample

File: snuba/datasets/querylog_processor.py
Comment (fpacifici) [inline snuba/datasets/querylog_processor.py]: there is a tuplify function in snuba.util

File: snuba/datasets/querylog_processor.py
Comment (fpacifici) [inline snuba/datasets/querylog_processor.py]: I think it would be worth to format this int oJSON instead of just serializing the mapping, Clickhosue has some json processing functions https://clickhouse.tech/docs/en/query_language/functions/json_functions/ that would make querying this dataset easier

File: snuba/datasets/querylog_processor.py
Comment (fpacifici) [inline snuba/datasets/querylog_processor.py]: Please ensure with @JTCunning  that the amount of rows inserted per second (from our metrics) will be ok or whether we need to sample.

**Review by fpacifici — CHANGES_REQUESTED** [review]
Please ensure you are able to consume and to query the dataset, also please let's get @JTCunning review on the DDL

File: snuba/datasets/querylog_processor.py
Comment (lynnagara) [reply snuba/datasets/querylog_processor.py]: I don't think it's doing the same thing as this

File: snuba/datasets/querylog_processor.py
Comment (lynnagara) [reply snuba/datasets/querylog_processor.py]: Confirmed that we don't want to sample here

File: snuba/datasets/querylog.py
Comment (fpacifici) [inline snuba/datasets/querylog.py]: I am not 100% sure about this field since there are datasets without org.

File: snuba/datasets/querylog.py
Comment (fpacifici) [inline snuba/datasets/querylog.py]: Is this the start or finish timestamp ? (probably better calling it accordingly so that we know which one it is)

File: snuba/datasets/querylog.py
Comment (fpacifici) [reply snuba/datasets/querylog.py]: ping

File: snuba/datasets/querylog_processor.py
Comment (fpacifici) [inline snuba/datasets/querylog_processor.py]: minor nit, I think you can inline this where you use it, it is one statement only and not even too long.

File: snuba/datasets/querylog_processor.py
Comment (fpacifici) [inline snuba/datasets/querylog_processor.py]: is this because the trace_id does not come in the right format ?

File: snuba/datasets/querylog_processor.py
Comment (fpacifici) [inline snuba/datasets/querylog_processor.py]: You should not need to transform the request_id to UUID and then back to string. If I remember correctly we had to do that for event_id and trace id because they did not come in the UUID format with the hyphens.  It should not be the case here.

File: snuba/request/schema.py
Comment (fpacifici) [inline snuba/request/schema.py]: This can potentially be non unique, though we would use this only for observability and with limited retention so I guess we are ok with it.

**Review by fpacifici — COMMENTED** [review]
This seems overall ok to me. Except for stats, I think it would be usefull to have the full stats payload in the dataset.
Please ensure the others in the team have a chance to see the table schema since we will have to use it for observability so we should ensure everyone can chime in on what we would like to have there. Specifically on whether we are ok changing the schema every time we want to add some additional information to track (I think that would be a serious drawback) or whether we will add some less typed data field like tags or similar.

File: snuba/datasets/querylog.py
Comment (lynnagara) [reply snuba/datasets/querylog.py]: Wouldn't the same technically apply to projects then? Since org ID is coming to errors/transactions and this is likely going to be a very useful to track down specfic customer issues I think it's worth putting it as a separate column

File: snuba/datasets/querylog.py
Comment (lynnagara) [reply snuba/datasets/querylog.py]: Are you suggesting adding some of those fields at a top level or having stats as a nested field? Personally I find it easier to write queries quickly without ARRAY JOIN (especially if we are already doing dealing with other nested fields in the dataset)

File: snuba/datasets/querylog_processor.py
Comment (lynnagara) [reply snuba/datasets/querylog_processor.py]: Thanks

File: snuba/datasets/querylog.py
Comment (lynnagara) [reply snuba/datasets/querylog.py]: Removed all the extensions and made this non queryable via API

File: snuba/datasets/querylog_processor.py
Comment (lynnagara) [reply snuba/datasets/querylog_processor.py]: Actually request id is getting the format without the hyphens here as well so this is needed

File: snuba/datasets/querylog_processor.py
Comment (lynnagara) [reply snuba/datasets/querylog_processor.py]: Yes, it doesn't have hyphens

File: snuba/datasets/querylog.py
Comment (lynnagara) [reply snuba/datasets/querylog.py]: Hmm, I matched this to "timestamp" in the timer (which I think means start)

File: snuba/datasets/querylog_processor.py
Comment (fpacifici) [reply snuba/datasets/querylog_processor.py]: how so ? I thought you generated it through uuid4()

Comment (lynnagara) [reply snuba/datasets/querylog_processor.py]: Well I did uuid4().hex when it's generated, just to keep things somewhat similar to the rest of our identifiers which I don't think use hyphens.

**Comment by lynnagara** [comment]
After thinking about stats some more, it seems like there is a lot of stuff in there of unkwown/ / variable value, it's probably better to dump it as a string and use JSON functions to extract the bits we want to query for.

File: snuba/datasets/querylog.py
Comment (fpacifici) [reply snuba/datasets/querylog.py]: After talking in person, in order to keep the dataset schema stable, it is better adding first the columns we need and not the stats payload as it is.

File: snuba/datasets/querylog.py
Comment (fpacifici) [inline snuba/datasets/querylog.py]: Actually after looking at commit https://github.com/getsentry/snuba/pull/783/commits/aee9d64c3a64c8db3ec2c77cb723618a8aa09f45, it does not seem bad to me. 
THere are a few fields we are going to query often. I would not remove them from the dataset as explicit columns whether you add or not a bare string with the whole stats.

Comment (lynnagara) [reply snuba/datasets/querylog.py]: Fine, I can add it back if you think it's useful. Yes, I think we need the whole stats as well for ad hoc queries.

Comment (lynnagara) [reply snuba/datasets/querylog.py]: Put those back now

File: snuba/datasets/querylog.py
Comment (JTCunning) [inline snuba/datasets/querylog.py]: Would it be worth making this a UInt8 and just multiplying the original value by 100?

File: snuba/datasets/querylog.py
Comment (JTCunning) [reply snuba/datasets/querylog.py]: I would be hesitant to add a sampling clause to this. It's an internal dataset. If we require sampling this data to get anything remotely accurate out of it, I believe we'd be doing something incorrect.

File: snuba/datasets/querylog.py
Comment (JTCunning) [inline snuba/datasets/querylog.py]: I don't personally think we need to sample, so if we decide not to, we can change this order to solely be `timestamp`.

**Review by JTCunning — COMMENTED** [review]
Schema looks great. I'll defer to others on the final +1.

File: snuba/datasets/querylog.py
Comment (lynnagara) [reply snuba/datasets/querylog.py]: Sample can be both float between 0-1 or integer values. UInt8 probably not large enough to cover all cases. I think it's more straightforward to preserve the original value just cast to float if necessary

File: snuba/datasets/querylog.py
Comment (fpacifici) [reply snuba/datasets/querylog.py]: ok

File: snuba/datasets/querylog.py
Comment (fpacifici) [reply snuba/datasets/querylog.py]: yeah I would still call it start_stimestamp so people don't have to try to understand how timer works to interpret the results.

File: snuba/datasets/querylog.py
Comment (fpacifici) [inline snuba/datasets/querylog.py]: mind adding the query_id from stats. That's the cache key for deduplication and cache. It may be useful to see how many of a given query run into cache/dedupe

File: snuba/datasets/querylog.py
Comment (fpacifici) [inline snuba/datasets/querylog.py]: please add the "consistent" field from stats, that tells us when we are sending the query to the first replica in the cluster.

File: snuba/datasets/querylog_processor.py
Comment (fpacifici) [inline snuba/datasets/querylog_processor.py]: Please add a comment and a TODO on why this is hardcoded to None now, otherwise, a few months down the road we will be yelling about this when we will figure out that the organizaiton column is always empty.

**Review by fpacifici — COMMENTED** [review]
This seems mostly ok to me, just a couple of nits, then please sync about the schema with James. It has to be applied to prod before we merge.
Also please ensure you run a local test with the real consumer to fill the real table. The consumer test only tests how the message is being processed, it does not try to add anything to Clickhouse, so it is not testing whether the INSERT statement you produce is actually working well with the schema.

**Comment by lynnagara** [comment]
Tested this e2e locally

**Review by fpacifici — COMMENTED** [review]
This is ok to me, please sync with James to get the table created in prod, then I will accept.

**Comment by lynnagara** [comment]
Updated with the storage changes and marked querylog as disabled so it is safe to merge.

File: snuba/datasets/factory.py
Comment (fpacifici) [inline snuba/datasets/factory.py:55]: If you are fixing the order, could you fix all of them ?

File: snuba/datasets/querylog_processor.py
Comment (fpacifici) [inline snuba/datasets/querylog_processor.py:51]: Mind verifying that these fields in stats are never None ? Because the default value `0` would be applied only if the keys are not present in the dictionary. If the key is present and None, this will try to add NULL to a non nullable column.

Actually applies to all the fields.

File: snuba/migrations/parse_schema.py
Comment (fpacifici) [inline snuba/migrations/parse_schema.py:106]: thanks

**Review by fpacifici — APPROVED** [review]
please ensure the settings change is done in production before merging

File: snuba/datasets/factory.py
Comment (lynnagara) [reply snuba/datasets/factory.py:55]: Sure

File: snuba/datasets/querylog_processor.py
Comment (lynnagara) [reply snuba/datasets/querylog_processor.py:51]: Good point, I updated to make sure `0` gets set even if the key is there and is none.

**Pull request merged**
